File: Product_Except_Itself.py
# // Time Complexity :O(n)
# // Space Complexity : O(1)
# // Did this code successfully run on Leetcode : yes
# // Three line explanation of solution in plain english

# create a result array and store prefix product while traverse
# from n to 1 traverse , calcualte postfix and multiply with the result on paticular index.


# // Your code here along with comments explaining your approach

class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:

        n = len(nums)
        result = [1] * n
        rp = 1

        for i in range(1, n) :
            rp = rp * nums[i-1]
            result[i] = rp
        rp = 1
        for i in range(n-1 , -1 , -1):
            result[i] *= rp
            rp *= nums[i]
        return result

File: ProductExceptItself.java
// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach

class Solution {

    public int[] productExceptSelf(int[] nums) {

        if(nums == null || nums.length == 0) return new int [0];
        int [] result = new int [nums.length];
        int rp = 1; int temp = 1;

        for(int i = 0; i < nums.length; i++){
            rp = rp * temp;
            result[i] = rp;
            temp = nums[i];
        }

        rp = 1; temp = 1;

        for(int i = nums.length - 1; i >=0; i--){
            rp = rp * temp;
            result[i] = result[i] * rp;
            temp = nums[i];
        }

        return result;

    }

}